lib/auth0/mixins/validation.rb

```diff
@@ -20,6 +28,118 @@ def validate_permissions_array(permissions)
         permissions.map { |permission| permission.to_h }
       end
 
+      class IdTokenValidator
+        def initialize(context)
+          @context = context
+        end
+
+        def validate(id_token)
+          begin
+            result = JWT.decode id_token, nil, false, {}
+          rescue JWT::DecodeError
+            raise Auth0::InvalidIdToken, 'ID token could not be decoded'
+          end
+
+          decoded_id_token = { payload: result.first, header: result.last, string: id_token }
```

Comment:
Yes, the [jwt](https://github.com/jwt/ruby-jwt) library performs the signature validation via the `decode` method, and to call `decode` I need the complete JWT string. There's no exposed API just for the signature validation AFAIK.

Reply:
Ok, then we need to refactor this because as of today it is decoding the same token twice. 

Rename the `validate_signature` to something like `validate_signature_and_decode` and make it return the payload claims dictionary/map that the `validate_claims` method would receive. As soon as the algorithm name is verified along with the signature, the header claims are no longer required.

Reply:
It's probably cleaner to leave it the way it is, since the decoding step checks for its own exceptions and conceptually decoding and signature validation are two distinct operations. The decoding is lightweight enough to not matter if it's being done twice.

Reply:
Changed in 64bc3aa057870bb29267f94fd0333c74b958ed82